core/templates/cowdata.h

```diff
@@ -272,139 +265,289 @@ void CowData<T>::_unref() {
+			const Error error = _realloc(new_size);
+			if (error) {
+				return error;
 			}
 		}
+
+		*_get_size() = new_size;
+	} else {
+		// Grow by forking.
+		// Use the max of capacity and new_size, to ensure we don't accidentally shrink after reserve.
+		const USize new_capacity = MAX(capacity(), (USize)new_size);
+		const Error error = _copy_to_new_buffer(new_capacity, size(), 1, 0);
+		if (error) {
+			return error;
+		}
 	}
 
-	// Set our new size.
-	*_get_size() = p_size;
+	// Create the new element at the given index.
+	memnew_placement(_ptr + new_size - 1, T(p_val));
 
 	return OK;
 }
 
+template <typename T>
+Error CowData<T>::reserve(USize p_min_capacity) {
+	ERR_FAIL_COND_V_MSG(p_min_capacity < (USize)size(), ERR_INVALID_PARAMETER, "reserve() called with a capacity smaller than the current size. This is likely a mistake.");
+
+	if (p_min_capacity <= capacity()) {
+		// No need to reserve more, we already have (at least) the right size.
+		return OK;
+	}
+
+	if (!_ptr) {
+		// Initial allocation.
+		return _alloc(p_min_capacity);
+	} else if (_get_refcount()->get() == 1) {
+		// Grow in-place.
+		return _realloc(p_min_capacity);
+	} else {
+		// Grow by forking.
+		return _copy_to_new_buffer(p_min_capacity, size(), 0, 0);
+	}
+}
+
 template <typename T>
 template <bool p_initialize>
 Error CowData<T>::resize(Size p_size) {
 	ERR_FAIL_COND_V(p_size < 0, ERR_INVALID_PARAMETER);
 
 	const Size prev_size = size();
 	if (p_size == prev_size) {
+		// Caller wants to stay the same size, so we do nothing.
 		return OK;
 	}
 
-	const Error error = _fork_allocate(p_size);
-	if (error) {
-		return error;
-	}
+	if (p_size > prev_size) {
+		// Caller wants to grow.
 
-	if constexpr (p_initialize) {
-		if (p_size > prev_size) {
+		if (!_ptr) {
+			// Grow by allocating.
+			const Error error = _alloc(p_size);
+			if (error) {
+				return error;
+			}
+		} else if (_get_refcount()->get() == 1) {
+			// Grow in-place.
+			if ((USize)p_size > capacity()) {
+				const Error error = _realloc(p_size);
+				if (error) {
+					return error;
+				}
+			}
+		} else {
+			// Grow by forking.
+			const Error error = _copy_to_new_buffer(p_size, prev_size, 0, 0);
+			if (error) {
+				return error;
+			}
+		}
+
+		// Construct new elements.
+		if constexpr (p_initialize) {
 			memnew_arr_placement(_ptr + prev_size, p_size - prev_size);
 		}
+		*_get_size() = p_size;
+
+		return OK;
 	} else {
-		static_assert(std::is_trivially_destructible_v<T>);
-	}
+		// Caller wants to shrink.
 
-	return OK;
+		if (p_size == 0) {
+			_unref();
+			return OK;
+		} else if (_get_refcount()->get() == 1) {
+			// Shrink in-place.
+			destruct_arr_placement(_ptr + p_size, prev_size - p_size);
+
+			// Shrink buffer if necessary.
+			const USize new_alloc_size = _get_alloc_size(p_size);
+			const USize prev_alloc_size = _get_alloc_size(capacity());
+			if (new_alloc_size < prev_alloc_size) {
+				_ALLOW_DISCARD_ _realloc_bytes(new_alloc_size); // Always succeeds on shrink.
```

Comment:
Not that I'm suggesting it is wise to do much alloc failure handling, but is there any guarantee that a realloc to a smaller size can't fail?
Grok suggests it is less likely than a grow, but it is not guaranteed.

If it can shrink by e.g. decommiting pages or another method that is one thing, but if it has to create a smaller buffer then copy, it's still allocating more than was there originally?